### PR TITLE
GH-129 Send level-up on same channel and remove after 5 seconds.

### PR DESCRIPTION
**Pull request by vLuckyyy**
Fixes GH-129

# Why?
Because to recude spam at the channels.

# Working

https://github.com/EternalCodeTeam/DiscordOfficer/assets/65517973/5f16985d-16ae-4fb8-8876-99ec22d5f99c

# READ-ME

Sure, you might wonder why the level-up message from a reaction is sent via DM. This is because, let's assume that a user received a level-up notification in the announcements channel; in that case, the level-up message would be sent on that channel, which is something we'd prefer to avoid. That's why I send the level-up message from the reaction via DM.

### CODE REVIEW

File: src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceMessageListener.java [hunk 1]
```diff
@@ -36,7 +35,8 @@ private void givePoints(MessageReceivedEvent event) {
         double points = (double) message.length / this.experienceConfig.messageExperience.howManyWords * basePoints;
         long userId = event.getAuthor().getIdLong();
 
-        this.experienceService.modifyPoints(userId, points, true).whenComplete((experience, throwable) -> {
+        long idLong = event.getChannel().getIdLong();
+        this.experienceService.modifyPoints(userId, points, true, idLong).whenComplete((experience, throwable) -> {
```
Comment (imDMK) [inline src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceMessageListener.java:39]:
```suggestion
        long channelId = event.getChannel().getIdLong();
        this.experienceService.modifyPoints(userId, points, true, channelId).whenComplete((experience, throwable) -> {
```

File: src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceReactionListener.java [hunk 2]
```diff
@@ -21,10 +23,16 @@ public void onMessageReactionAdd(MessageReactionAddEvent event) {
...
 
-        this.experienceService.modifyPoints(userId, points, true).whenComplete((experience, throwable) -> {
-            if (throwable != null) {
-                throwable.printStackTrace();
-            }
+        if (event.getUser().isBot()) {
+            return;
+        }
+
+        event.getUser().openPrivateChannel().queue(channel -> {
```
Comment (imDMK) [inline src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceReactionListener.java]:
```suggestion
        User user = event.getUser();
        
        if (user.isBot()) {
            return;
        }

        user.openPrivateChannel().queue(channel -> {
```

File: src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceReactionListener.java [hunk 3]
```diff
@@ -33,10 +41,16 @@ public void onMessageReactionRemove(MessageReactionRemoveEvent event) {
...
```
Comment (imDMK) [inline src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceReactionListener.java]:
```suggestion
        User user = event.getUser();
        
        if (user.isBot()) {
            return;
        }

        user.openPrivateChannel().queue(channel -> {
```

File: src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceVoiceListener.java [hunk 4]
```diff
@@ -51,7 +51,9 @@ private void leaveVoiceChannel(GuildVoiceUpdateEvent event) {
 
         long userId = event.getMember().getIdLong();
         this.usersVoiceActivityData.usersOnVoiceChannel.remove(event.getMember().getIdLong());
-        this.experienceService.modifyPoints(userId, this.calculatePoints(event), true).whenComplete((experience, throwable) -> {
+
+        long idLong = event.getChannelLeft().getIdLong();
+        this.experienceService.modifyPoints(userId, this.calculatePoints(event), true, idLong).whenComplete((experience, throwable) -> {
```
Comment (imDMK) [inline src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceVoiceListener.java:56]:
```suggestion
        long channelIdLeft = event.getChannelLeft().getIdLong();
        this.experienceService.modifyPoints(userId, this.calculatePoints(event), true, channelIfLeft).whenComplete((experience, throwable) -> {
```

File: src/main/java/com/eternalcode/discordapp/leveling/games/CodeGameAnswerController.java [hunk 5]
```diff
@@ -66,7 +66,7 @@ public void onMessageReceived(MessageReceivedEvent event) {
                 .setDescription(formatter.format(this.codeGameConfiguration.embedSettings.description))
                 .setFooter(this.codeGameConfiguration.embedSettings.footer);
 
-            this.experienceService.modifyPoints(event.getAuthor().getIdLong(), points, true)
+            this.experienceService.modifyPoints(event.getAuthor().getIdLong(), points, true, event.getChannel().getIdLong())
```
Comment (imDMK) [inline src/main/java/com/eternalcode/discordapp/leveling/games/CodeGameAnswerController.java]: From what I see, `event.getChannel()` is already used twice in this class
Store a variable.

File: src/main/java/com/eternalcode/discordapp/leveling/LevelController.java [hunk 6]
```diff
@@ -73,9 +80,21 @@ public void update(ExperienceChangeEvent experienceChangeEvent) {
...
                 .format(this.levelConfig.message.description);
 
-            TextChannel levelChannel = this.jda.getTextChannelById(this.levelConfig.channel);
-            if (levelChannel != null) {
-                levelChannel.sendMessage(messageContent).queue();
+
+            try {
+                Optional<MessageChannel> textChannelOptional = Optional.ofNullable(this.jda.getTextChannelById(event.channelId()));
+
+                MessageChannel channel = textChannelOptional.orElse(this.jda.getPrivateChannelById(event.channelId()));
```
Comment (imDMK) [inline src/main/java/com/eternalcode/discordapp/leveling/LevelController.java:87]: Create a separate method that checks whether the channel exists and methods for acquiring channels.
```suggestion
                long channelId = event.getChannelId();

                MessageChannel channel = this.isPrivateChannel(channelId) ? this.getPrivateChannel(channelId) : this.getTextChannel(channelId);
```

**Review by imDMK — CHANGES_REQUESTED** [review]

**Review by imDMK — APPROVED** [review]

File: src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceReactionListener.java [hunk 7]
```diff
@@ -21,23 +22,38 @@ public void onMessageReactionAdd(MessageReactionAddEvent event) {
...
+            return;
+        }
+
+        this.modifyPoints(user, userId, points);
+    }
+
+    private void modifyPoints(User event, long userId, double points) {
+        long channel = event.openPrivateChannel().complete().getIdLong();
+
+        this.experienceService.modifyPoints(userId, points, true, () -> channel)
```
Comment (Rollczi) [inline src/main/java/com/eternalcode/discordapp/leveling/experience/listener/ExperienceReactionListener.java:51]:
```suggestion
        LongSupplier channel = () -> event.openPrivateChannel().complete().getIdLong();

        this.experienceService.modifyPoints(userId, points, true, channel)
```

**Review by Rollczi — APPROVED** [review]

**Pull request merged**